include/openenclave/internal/atomic.h

```diff
@@ -38,4 +42,36 @@ OE_INLINE uint64_t oe_atomic_decrement(volatile uint64_t* x)
 #endif
 }
 
+OE_INLINE
+bool oe_atomic_compare_and_swap(
+    int64_t volatile* dest,
+    int64_t old,
+    int64_t newval)
+{
+#if defined(__GNUC__)
+    return __atomic_compare_exchange_n(
+        dest, &old, newval, 1, __ATOMIC_ACQ_REL, __ATOMIC_ACQUIRE);
```

Comment:
Maybe it is better to use __ATOMIC_ACQ_REL to provide both acquire and release semantics?
Otherwise from the function name `oe_atomic_compare_and_swap` it is not clear that it provides only acquire semantics.

Reply:
The memory model for the success case is indeed __ATOMIC_ACQ_REL.

The last parameter is for memory model of the failure case when the exchange doesn't happen. The gcc document says:

> This memory model cannot be __ATOMIC_RELEASE nor __ATOMIC_ACQ_REL. It also cannot be a stronger model than that specified by success_memmodel.